go/stats/pull_backend.go

```diff
@@ -0,0 +1,127 @@
+	"sync"
+
+	log "github.com/golang/glog"
+)
+
+type PullBackend interface {
+	NewMultiCounter(*MultiCounters, string)
+	NewMultiCounterFunc(*MultiCountersFunc, string)
+	NewCounter(*Counters, string)
+	// Add more functions here that need to be registered with Prom
+}
+
+var (
+	be = struct {
+		sync.RWMutex
+		m map[string]PullBackend
+	}{m: make(map[string]PullBackend)}
+
+	pullMultiCounters = struct {
+		sync.RWMutex
+		m map[string]*MultiCounters
+	}{m: make(map[string]*MultiCounters)}
+
+	pullCounters = struct {
+		sync.RWMutex
+		m map[string]*Counters
+	}{m: make(map[string]*Counters)}
+
+	pullMultiCountersFunc = struct {
+		sync.RWMutex
+		m map[string]*MultiCountersFunc
+	}{m: make(map[string]*MultiCountersFunc)}
+)
+
+func publishPullMultiCounters(mc *MultiCounters, name string) {
+	pullMultiCounters.Lock()
+	pullMultiCounters.m[name] = mc
+	pullMultiCounters.Unlock()
+
+	be.RLock()
+	defer be.RUnlock()
+	for _, backend := range be.m {
+		backend.NewMultiCounter(mc, name)
+	}
+}
+
+func flushMultiCounters() {
+	// Is this the right locking model? Could maybe copy pmc here instead
+	pullMultiCounters.RLock()
+	defer pullMultiCounters.RUnlock()
+	for name, mc := range pullMultiCounters.m {
+		be.RLock()
+		for _, backend := range be.m {
+			backend.NewMultiCounter(mc, name)
+		}
+		be.RUnlock()
+	}
+}
+func GetMultiCounter(name string) map[string]int64 {
+	pullMultiCounters.RLock()
+	defer pullMultiCounters.RUnlock()
+	return pullMultiCounters.m[name].Counters.Counts()
+}
+
+func publishPullCounters(c *Counters, name string) {
+	pullCounters.Lock()
+	pullCounters.m[name] = c
+	pullCounters.Unlock()
+
+	be.RLock()
+	defer be.RUnlock()
+	for _, backend := range be.m {
+		backend.NewCounter(c, name)
+	}
+}
+
+func flushCounters() {
+	pullCounters.RLock()
+	defer pullCounters.RUnlock()
+	for name, c := range pullCounters.m {
+		be.RLock()
+		for _, backend := range be.m {
+			backend.NewCounter(c, name)
+		}
+		be.RUnlock()
+	}
+}
+
+func GetCounter(name string) map[string]int64 {
+	pullCounters.RLock()
+	defer pullCounters.RUnlock()
+	return pullCounters.m[name].Counts()
+}
+
+func publishPullMultiCountersFunc(mcf *MultiCountersFunc, name string) {
+	pullMultiCountersFunc.Lock()
+	pullMultiCountersFunc.m[name] = mcf
+	pullMultiCountersFunc.Unlock()
+
+	be.RLock()
+	defer be.RUnlock()
+	for _, backend := range be.m {
+		backend.NewMultiCounterFunc(mcf, name)
+	}
+}
+
+func GetMultiCounterFunc(name string) map[string]int64 {
+	pullMultiCountersFunc.RLock()
+	defer pullMultiCountersFunc.RUnlock()
+	// Add error handling? Should we call the func from the collect itself?
+	return pullMultiCountersFunc.m[name].CountersFunc.Counts()
+}
+
+// RegisterPullBackendImpl registers an implementation of PullBackend
+func RegisterPullBackendImpl(name string, pullBackend PullBackend) {
+	be.Lock()
+	if _, ok := be.m[name]; ok {
+		log.Fatalf("PullBackend named %v already exists", name)
+	}
+	be.m[name] = pullBackend
+	be.Unlock()
+	flushMultiCounters()
```

Comment:
Since flush is only ever used in this call site, I feel like you should just inline the loop to make it clearer.